### PR TITLE
bind: Expose methods for intercepting transactions

### PR DESCRIPTION
**Pull request by jmcnevin**
On occasion, we have needed to access the BoundContract or ABI objects inside of a contract wrapper, but since `contract` was not an exported field, we have had to do extra legwork, essentially building up our own bindings from scratch, which isn't fun.

The PR allows users of abigen-created libraries lower-level access to the transaction lifecycle, and the ABI byte packings, with minimal changes to the code that abigen is creating (the *Raw types are made redundant by these changes).

### CODE REVIEW

**Comment by jmcnevin** [comment]
@karalabe Not sure who would be best to get eyes on this

File: accounts/abi/bind/base.go [hunk 1]
```diff
@@ -180,9 +180,13 @@ func (c *BoundContract) Transfer(opts *TransactOpts) (*types.Transaction, error)
 	return c.transact(opts, &c.address, nil)
 }
 
-// transact executes an actual transaction invocation, first deriving any missing
-// authorization fields, and then scheduling the transaction for execution.
-func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, input []byte) (*types.Transaction, error) {
+// RawInput returns the raw bytes encoding a method call and its provided arguments
+func (c *BoundContract) RawInput(method string, params ...interface{}) ([]byte, error) {
+	return c.abi.Pack(method, params...)
```
Comment (gballet) [inline accounts/abi/bind/base.go]: I would prefer `GetRawInput`.

File: accounts/abi/bind/base.go [hunk 2]
```diff
@@ -180,9 +180,13 @@ func (c *BoundContract) Transfer(opts *TransactOpts) (*types.Transaction, error)
...
+}
+
+// RawTransaction creates an unsigned transaction, but does not sign or send the transaction
+func (c *BoundContract) RawTransaction(opts *TransactOpts, contract *common.Address, input []byte) (*types.Transaction, error) {
```
Comment (gballet) [inline accounts/abi/bind/base.go]: This should be called `NewRawTransaction`

File: accounts/abi/bind/template.go [hunk 3]
```diff
@@ -106,7 +106,7 @@ var (
 		  if err != nil {
 		    return common.Address{}, nil, nil, err
 		  }
-		  return address, tx, &{{.Type}}{ {{.Type}}Caller: {{.Type}}Caller{contract: contract}, {{.Type}}Transactor: {{.Type}}Transactor{contract: contract}, {{.Type}}Filterer: {{.Type}}Filterer{contract: contract} }, nil
```
Comment (gballet) [inline accounts/abi/bind/template.go]: Is there a reason for renaming `contract` to `BoundContract`? I'm not a big fan of this, since people might regenerate their bindings in order to bring in newer features, and then they potentially have to rewrite their entire codebase.

File: accounts/abi/bind/template.go [hunk 4]
```diff
@@ -154,28 +154,13 @@ var (
 	  TransactOpts bind.TransactOpts    // Transaction auth options to use throughout this session
 	}
 
-	// {{.Type}}Raw is an auto generated low-level Go binding around an Ethereum contract.
-	type {{.Type}}Raw struct {
```
Comment (gballet) [inline accounts/abi/bind/template.go]: You are removing something that doesn't seem to be replaced anywhere else, and it beaks the test. Is that change intentional ?

File: accounts/abi/bind/template.go [hunk 5]
```diff
@@ -214,44 +199,6 @@ var (
 	  return bind.NewBoundContract(address, parsed, caller, transactor, filterer), nil
 	}
 
-	// Call invokes the (constant) contract method with params as input values and
```
Comment (gballet) [inline accounts/abi/bind/template.go]: Same question as above

File: accounts/abi/bind/base.go [hunk 6]
```diff
@@ -231,13 +235,33 @@ func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, i
 	} else {
 		rawTx = types.NewTransaction(nonce, c.address, value, gasLimit, gasPrice, input)
 	}
+
+	return rawTx, nil
```
Comment (CoreyLin) [inline accounts/abi/bind/base.go]: I can't comment on line 231. It said "// Create the transaction, sign it and schedule it for execution".

Here since this method only creates the rawTx, suggest to modify it to "Create the raw transaction"

**Comment by jmcnevin** [comment]
I won't have any further time to work on this, so closing for now.
